### DISCUSSION TITLE
Shifts the values of the record to the right by one

### DISCUSSION THREAD

**Discussion by aborruso** [johnkerl/miller#1144] [Q&A] [answered]
Hi,
I'm sure it's very easy to do it.

I have this no header CSV

```
a,2,3,
2,3,,
b,4,5,
```

And I want to transform it, shifting all values to the right by one, from this (I am adding the header also to make my target more readable)

| 1 | 2 | 3 | 4 |
| --- | --- | --- | --- |
| a | 2 | 3 |  |
| 2 | 3 |  |  |
| b | 4 | 5 |  |

to this one

| 1 | 2 | 3 | 4 |
| --- | --- | --- | --- |
|  | a | 2 | 3 |
|  | 2 | 3 |  |
|  | b | 4 | 5 |

I thought of using positional field names in this way

```
mlr --csv --implicit-csv-header put 'for (k in $*){$[ $[[k+1]] ]=$[ $[[k]] ]}' input.csv
```

but I get back

```
Record/map indices must be string, int, or array thereof; got error
```

And I'm sure there will be a very easy array way to do it.
I'm not able to do it.

Thank you

**Comment by aborruso** [comment]
And my way has no sense, also correcting it.

Running

```
mlr --csv --implicit-csv-header put  'for (k in $*){$[[[int(k)+1]]]=$[[[int(k)]]]}' input.csv
```

I have all equal to the first field

| 1 | 2 | 3 | 4 |
| --- | --- | --- | --- |
| a | a | a | a |
| 2 | 2 | 2 | 2 |
| b | b | b | b |

**Comment by johnkerl** [comment, marked as answer]
@aborruso how about:

```
$ cat shifter.mlr
keys   = get_keys($*);
values = get_values($*);

newrec = {};
newrec[keys[1]] = "";
for (i = 1; i <= length(keys); i += 1) {
    newrec[keys[i+1]] = values[i];
}

$* = newrec;
```
```
$ mlr --implicit-csv-header --c2p --from input.csv cat
1 2 3 4
a 2 3 -
2 3 - -
b 4 5 -
```

```
$ mlr --implicit-csv-header --c2p --from input.csv put -f shifter.mlr
1 2 3 4
- a 2 3
- 2 3 -
- b 4 5
```

**Reply by johnkerl** [reply]
Or this:

```
keys   = get_keys($*);
values = get_values($*);

newrec = {};
# newrec[keys[1]] = "";
newrec[keys[1]] = values[length(keys)]; # <----- here
for (i = 1; i <= length(keys); i += 1) {
    newrec[keys[i+1]] = values[i];
}

$* = newrec;
```

will "rotate" the last column's values into the first column

**Reply by aborruso** [reply]
Thank you very much, great.

`shift` could be a new verb :)
